enhancements/microshift/microshift-updateability-ostree.md

```diff
@@ -0,0 +1,944 @@
+runs (for alternative see [Performing backup on shutdown](#performing-backup-on-shutdown))
+and with MicroShift cluster (i.e. etcd, kube-apiserver, etc.) not being off.
+As a general rule: if previous boot was healthy, data will be backed up, if boot
+was unhealthy, data will be restored.
+If system rolled back, we want compatible and healthy data to be restored,
+but, if previous commit was healthy, after backing up its data.
+
+To provide ability to rollback and restore MicroShift data, backup for each
+ostree commit/deployment will be kept.
+
+For now, default behavior will be to keep backups even if the commit no longer
+exists on the system. Reason for this is that reintroducing commit results in
+the same ID (it is a checksum) and deleting backups would prevent restoring
+healthy data for the commit. This could be a configurable option, but rules of
+pruning old backups are out of scope of this enhancement.
+We will need to provide documentation on how can admin remove old backups.
+
+It is worth noting that although enhancement focuses on MicroShift's data,
+backups will be tied to specific ostree commits.
+Linking backups to ostree commits will ensure that staging and rolling back
+is "all or nothing" and MicroShift does not accidentally run applications
+belonging to different commits (especially different might not be MicroShift
+itself, but the applications that run on top of it).
+Also see [Deciding to backup or restore based on MicroShift health, rather than system's](#deciding-to-backup-or-restore-based-on-microshift-health-rather-than-systems)
+alternative.
+
+### Integration with greenboot
+
+greenboot integrates with systemd, ostree, and grub to provide auto-healing
+capabilities of newly staged and booted deployment in form of reboot:
+if system is still unhealthy after specified amount of reboots, it
+will be rolled back to previous ostree deployment (commit).
+
+Because greenboot already exists as an integral part of Red Hat Device Edge
+systems, we will integrate with it, rather than creating a new system.
+
+greenboot determines system health using health check scripts and MicroShift
+already provides such script. For more information about greenboot and existing
+integration see [Integrating MicroShift with Greenboot](./microshift-greenboot.md).
+It is worth noting that MicroShift health check for greenboot is part of MicroShift
+RPMs, so in case of a bug MicroShift team is able to provide fix with next
+release, and health checks do not have to be compatible with other versions of MicroShift.
+
+After health check, either "green" (system is healthy) or "red" (system is unhealthy)
+scripts are executed. MicroShift will provide "green" and "red" scripts which
+will persist the overall system's health for the current ostree commit.
+
+Due to how ostree systems are updated (booted into new root filesystem),
+MicroShift cannot prevent users from following unsupported upgrade paths.
```

Comment:
That's not strictly true.  It's absolutely possible to have client side logic which verifies state before reboot.  Of course, there's a bit of a [GIGO](https://en.wikipedia.org/wiki/Garbage_in,_garbage_out) problem here in that we'd be working around whatever is providing input to the system for the upgrade target not providing the correct thing.

This also relates to https://github.com/ostreedev/ostree/issues/2855

Reply:
Nice, it looks like that's still WIP so it may be an area for us to enhance in the future.